File: leetcode/swift/p58-length-of-last-word.swift
// LeetCode 58. Length of Last Word.

class Solution {
    func lengthOfLastWord(_ s: String) -> Int {
        if let lastWord = s.split(separator: " ").last {
            return lastWord.count
        }
        return 0
    }
}

struct TestCase {
    var s : String
    var exp : Int
}

let test_cases = [
    TestCase(s: "", exp: 0),
    TestCase(s: "a", exp: 1),
    TestCase(s: "a ", exp: 1),
    TestCase(s: " a", exp: 1),
    TestCase(s: "Hello World", exp: 5),
    TestCase(s: "   fly me   to   the moon  ", exp: 4),
    TestCase(s: "luffy is still joyboy", exp: 6),
]

var solution = Solution()
for tc in test_cases {
    let act = solution.lengthOfLastWord(tc.s)
    if tc.exp != act {
        print("Fail. s: '\(tc.s)', exp: \(tc.exp), act: \(act)")
    }
}

File: leetcode/rust/src/p58-length-of-last-word.rs
struct Solution {}

impl Solution {
    pub fn length_of_last_word(s: String) -> i32 {
        let s2 = s.trim_end();
        match s2.rfind(' ') {
            Some(i) => return (s2.len() - i - 1) as i32,
            None => return s2.len() as i32,
        }
    }
}

fn main() {
    struct TestCase {
        s: String,
        exp: i32,
    }
    let test_cases = vec![
        TestCase {
            s: "".to_string(),
            exp: 0,
        },
        TestCase {
            s: "a".to_string(),
            exp: 1,
        },
        TestCase {
            s: "a ".to_string(),
            exp: 1,
        },
        TestCase {
            s: " a".to_string(),
            exp: 1,
        },
        TestCase {
            s: "Hello World".to_string(),
            exp: 5,
        },
        TestCase {
            s: "   Hello World   ".to_string(),
            exp: 5,
        },
        TestCase {
            s: "there".to_string(),
            exp: 5,
        },
    ];

    for tc in test_cases.iter() {
        let act = Solution::length_of_last_word(tc.s.clone());
        if tc.exp != act {
            println!("Fail, s: {:?}, exp: {:?}, act: {:?}", tc.s, tc.exp, act);
        }
    }
}
